ext/liquid_c/tag_markup.c

```diff
@@ -0,0 +1,73 @@
+#include "liquid.h"
+#include "tag_markup.h"
+
+static VALUE cLiquidCTagMarkup;
+
+static void tag_markup_mark(void *ptr)
+{
+    tag_markup_t *markup = ptr;
+
+    rb_gc_mark(markup->markup);
+    rb_gc_mark(markup->tag_name);
+    rb_gc_mark(markup->block_body_obj);
+}
+
+static void tag_markup_free(void *ptr)
+{
+    xfree(ptr);
+}
+
+static size_t tag_markup_memsize(const void *ptr)
+{
+    return sizeof(tag_markup_t);
+}
+
+const rb_data_type_t tag_markup_data_type = {
+    "liquid_tag_markup",
+    { tag_markup_mark, tag_markup_free, tag_markup_memsize, },
+    NULL, NULL, RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+VALUE tag_markup_new(VALUE tag_name, VALUE markup, bool unknown)
+{
+    tag_markup_t *tag;
+    VALUE obj = TypedData_Make_Struct(cLiquidCTagMarkup, tag_markup_t, &tag_markup_data_type, tag);
+
+    tag->flags = 0;
+    if (unknown) tag->flags |= TAG_FLAG_UNKNOWN;
+    tag->tag_name = tag_name;
+    tag->markup = markup;
+    tag->block_body = NULL;
+
+    return obj;
+}
+
+VALUE tag_markup_get_tag_name(VALUE self)
+{
+    tag_markup_t *tag;
+    TagMarkup_Get_Struct(self, tag);
+    return tag->tag_name;
+}
+
+VALUE tag_markup_get_markup(VALUE self)
+{
+    tag_markup_t *tag;
+    TagMarkup_Get_Struct(self, tag);
+    return tag->markup;
+}
+
+void tag_markup_set_block_body(VALUE self, VALUE block_body_obj, block_body_t *block_body)
+{
+    tag_markup_t *tag;
+    TagMarkup_Get_Struct(self, tag);
+    assert(tag->block_body == NULL);
+    tag->block_body_obj = block_body_obj;
+    tag->block_body = block_body;
+}
+
+void liquid_define_tag_markup()
+{
+    cLiquidCTagMarkup = rb_define_class_under(mLiquidC, "TagMarkup", rb_cObject);
+    rb_global_variable(&cLiquidCTagMarkup);
```

Comment:
Unrelated to this PR, but why is this needed? Aren't all constants global?
Is it to prevent GC? I'm not seeing this line being used with `rb_define_class_under` in Ruby's code.

Reply:
It seems like `rb_global_variable` adds the variable to a global roots list that is pinned during GC which prevents the compactor from moving it.